### DISCUSSION TITLE
Indexing takes a long time on highest Cloud RavenDB tier

### DISCUSSION THREAD

**Discussion by florisrobbemont** [ravendb/ravendb#14197] [Q&A]
Hi everyone,

We are currently developing a time-reporting solution that stores data in RavenDB. We use a cloud hosted version (using RavenDB Cloud) and are currently on the highest production tier (P50).

Compared to my local machine I would've expected to see higher indexing speeds, but the overal speed seems to hang around 3000 docs per second (compared to local where I see speeds up to 20K docs per second). My local machine doesn't have 32 Cores and 128GB RAM.

I'm sure we're doing something wrong, or missing some important bits. Maybe it's the design of our documents?

Does anyone have any tips for us? Maybe someone has experienced the same using the cloud version?

For example, the following index has currently dropped to around 1000 docs per second (on P50 instance that isn't used at all). The size of the collection is 8 million documents.

![image](https://user-images.githubusercontent.com/1893978/167869335-493dc705-7bfa-40ab-beae-c090a5198d0a.png)

The documents look like:

<img width="896" alt="image" src="https://user-images.githubusercontent.com/1893978/167869615-2d8a37f5-e8d0-491a-9a13-103307ade123.png">

Looking forward to seeing if anyone else is experiencing this.

Cheers,
Floris

**Comment by ayende** [comment]
How much time are we talking about?
How much data is involved?

Please note that for indexing - the disk type you have is typically far more relevant. What disk are you using?

**Reply by florisrobbemont** [reply]
Wow quick answer! Thnx!

The current run (rebuild index) I'm trying is already taking over an hour and is at 85% now. We also have more complex indexes and documents in this solution which take even longer (longest we've seen a complete rebuild take is over 40 hours).

This collection (UserBankLogs) is around 8 million documents. We're using the premium storage option for this one:
![image](https://user-images.githubusercontent.com/1893978/167871700-af6a793a-b231-4f2f-bf66-bd361955244d.png)

**Reply by ayende** [reply]
Note that it would probably be better to run those indexes in parallel, you have the capacity for this.
Can you provide the details here:

![image](https://user-images.githubusercontent.com/116915/167872691-66af7d12-88f5-4506-ba63-ba49ba8aaaf0.png)

**Reply by florisrobbemont** [reply]
How can I run these indexes in parallel? The deployment mode is already in parallel I believe.
Attached is a portion of the indexing performance (I just started a new rebuild). If it's not enough data I can re-upload a bigger one later today
[indexPerf of ClockingService 2022-05-11 16-37.txt](https://github.com/ravendb/ravendb/files/8670844/indexPerf.of.ClockingService.2022-05-11.16-37.txt)
.

**Reply by ayende** [reply]
are you creating the indexes one at a time?

**Reply by ayende** [reply]
Note that you are indexing at a rate of ~1K / sec - and the limit seems to be the batch size

![image](https://user-images.githubusercontent.com/116915/167880349-8a5ffd41-7367-42ed-a920-43b53c0284bc.png)

You can increase this for the index using the `Indexing.MapBatchSize` configuration item.

**Reply by florisrobbemont** [reply]
In production the indexes are already there. When we push a new version it needs to rebuild, and this is where it get's slow (last rebuild of the index took 40+ hours and consists of 35M documents (2 collections, index size is 15 million). Not sure if this answers your question?

The batch size is something I've tried earlier, but even at 128K batch size the performance dropped to under 1K.

**Reply by florisrobbemont** [reply]
BTW, not sure if this is relevant, but the Indexing.MapBatchSize wasn't set on this instance (left to default). Which according the docs should mean there's no batch limit. Is there a different setting that locks it down to 32K?

**Reply by ayende** [reply]
Did you set it on the index directly, maybe? 
If you want, you can create a support ticket from the cloud portal, we can then look at the details on the instance directly.

**Reply by arekpalinski** [reply]
@florisrobbemont The indexing batch limit is set by default in RavenDB Cloud and it depends on the number of IOPS that the machine's storage has. The reasoning is that we found many times that people tent to forget about disk perf and overwhelming the instance with high number of indexes - especially when doing initial deployments so the indexes needs to be built from scratch. Over the time once you have a regular load then it's less important.

Now you have 2300 IOPS but I see that you had 5000 IOPS configured at the time your opened this discussion. If you moved to 7500 IOPS then the limit would be automatically adjusted to size size of 64K docs per batch. Note that's the default setting which you can override per index (in the index configuration) or per database (in the database settings) via the RavenDB Studio in order to adjust it to your requirements. We put the limit based on the IOPS number and so far it did a good job although it makes a difference if someone will put 10 indexes vs 1000 indexes. What's why it's still up to user to configure it.

**Reply by florisrobbemont** [reply]
@arekpalinski Thnx for the info! We'll take a look at the batch size of individual indexes, maybe we can fine tune it a bit. I guess we're a little taken back by the performance on the highest tier, especially when compared to locally. Maybe we're modeling our documents wrong?

What I did see was the speed of the indexing process slowing down towards 60% progress. We've even seen indexing at around 150 docs/sec in some more complex indexes. In normal production this isn't a big issue ofcourse, but when we need to rebuild because of index changes waiting 40 hours will become an issue.

Are there any guidelines as to how high we can set these batches? At which point is there a diminishing return? 

I'm just looking for tips on how to tune the system. We've migrated from a SQL database and are currently not getting the performance from RavenDB that we expected. Any help is appreciated 🙌

**Reply by arekpalinski** [reply]
In order to speed up the processing of an updated index definition the best would be to change the performance tier of your storage for the duration of the indexing.  Regarding the size of batches, if you don't have many indexes (e.g. 10 - 20 per database) and a few databases and you are on P50 instance, then I'd say that even dropping the batch size limit is reasonable.

When it comes to the analysis of your models and indexes I'll contact you privately for that.

**Reply by florisrobbemont** [reply]
@arekpalinski We have 4 databases on the entire instance, and only 12 indexes on this specific database. Our biggest concern at the moment is that our data is not that complex (simple documents under 1KB), our indexes are not that complex (simple map reduces), but it still takes a very long time to process the data.

You can contact me by e-mail if you want 😊 It should be in my Github profile.

**Reply by arekpalinski** [reply]
I already emailed you. 

The problem I noticed in the indexing perf you attached yesterday is that we don't actually know where RavenDB spent most of the time in each indexing batch for that index:

![image](https://user-images.githubusercontent.com/376551/168090183-c60d1d6d-1774-4b08-a655-a1791d7cd288.png)

The bar at the bottom is supposed to fill most of the one above - `UserBankLogs (Collection)`. But in your case average batch took 30 seconds while the stages we measured was only about 2 seconds. The questions is where did we spent the rest of the time.

![image](https://user-images.githubusercontent.com/376551/168090427-5ab12ab7-b820-46b1-b769-faeca2d5a7b3.png)

**Reply by florisrobbemont** [reply]
Here's another export with some more index performance. It seems the same is happening with this index, but the other index isn't having the same gap. Hopefully this helps.

[indexPerf of ClockingService 2022-05-12 16-59.txt](https://github.com/ravendb/ravendb/files/8680282/indexPerf.of.ClockingService.2022-05-12.16-59.txt)

**Comment by arekpalinski** [comment]
The investigation of the issue resulted in the following PR: https://github.com/ravendb/ravendb/pull/14229
